File: apps/api/internal/domain/entity/specification.go
package entity

import (
	"time"

	"github.com/google/uuid"
)

type SpecStatus string

const (
	SpecStatusDraft     SpecStatus = "draft"
	SpecStatusGenerated SpecStatus = "generated"
	SpecStatusReviewed  SpecStatus = "reviewed"
	SpecStatusApproved  SpecStatus = "approved"
	SpecStatusRejected  SpecStatus = "rejected"
)

type Specification struct {
	ID                    string              `bson:"_id" json:"id"`
	ProjectID             string              `bson:"project_id" json:"project_id"`
	Version               int                 `bson:"version" json:"version"`
	Status                SpecStatus          `bson:"status" json:"status"`
	Overview              string              `bson:"overview" json:"overview"`
	Objectives            []string            `bson:"objectives" json:"objectives"`
	FeatureBreakdown      []FeatureSpec       `bson:"feature_breakdown" json:"feature_breakdown"`
	UserRolesPermissions  []RolePermission    `bson:"user_roles_permissions" json:"user_roles_permissions"`
	TechnicalArchitecture TechArchitecture    `bson:"technical_architecture" json:"technical_architecture"`
	IntegrationReqs       []IntegrationReq    `bson:"integration_requirements" json:"integration_requirements"`
	TimelineEstimate      TimelineEstimate    `bson:"timeline_estimate" json:"timeline_estimate"`
	CostEstimate          CostEstimate        `bson:"cost_estimate" json:"cost_estimate"`
	Assumptions           []string            `bson:"assumptions" json:"assumptions"`
	Risks                 []Risk              `bson:"risks" json:"risks"`
	InternalNotes         []InternalNote      `bson:"internal_notes,omitempty" json:"internal_notes,omitempty"`
	ApprovedBy            string              `bson:"approved_by,omitempty" json:"approved_by,omitempty"`
	ApprovedAt            time.Time           `bson:"approved_at,omitempty" json:"approved_at,omitempty"`
	CreatedAt             time.Time           `bson:"created_at" json:"created_at"`
	UpdatedAt             time.Time           `bson:"updated_at" json:"updated_at"`
}

type FeatureSpec struct {
	Name         string   `bson:"name" json:"name"`
	Description  string   `bson:"description" json:"description"`
	SubFeatures  []string `bson:"sub_features" json:"sub_features"`
	Priority     string   `bson:"priority" json:"priority"`
	EstimatedHrs float64  `bson:"estimated_hours" json:"estimated_hours"`
}

type RolePermission struct {
	Role        string   `bson:"role" json:"role"`
	Permissions []string `bson:"permissions" json:"permissions"`
}

type TechArchitecture struct {
	Frontend       []string `bson:"frontend" json:"frontend"`
	Backend        []string `bson:"backend" json:"backend"`
	Database       []string `bson:"database" json:"database"`
	Infrastructure []string `bson:"infrastructure" json:"infrastructure"`
	Suggestions    string   `bson:"suggestions,omitempty" json:"suggestions,omitempty"`
}

type IntegrationReq struct {
	Name        string `bson:"name" json:"name"`
	Type        string `bson:"type" json:"type"`
	Description string `bson:"description" json:"description"`
}

type TimelineEstimate struct {
	TotalWeeks int             `bson:"total_weeks" json:"total_weeks"`
	Phases     []PhaseEstimate `bson:"phases" json:"phases"`
}

type PhaseEstimate struct {
	Name     string `bson:"name" json:"name"`
	Weeks    int    `bson:"weeks" json:"weeks"`
	Details  string `bson:"details,omitempty" json:"details,omitempty"`
}

type CostEstimate struct {
	TotalMin  float64         `bson:"total_min" json:"total_min"`
	TotalMax  float64         `bson:"total_max" json:"total_max"`
	Currency  string          `bson:"currency" json:"currency"`
	Breakdown []CostBreakdown `bson:"breakdown" json:"breakdown"`
}

type CostBreakdown struct {
	Category string  `bson:"category" json:"category"`
	Min      float64 `bson:"min" json:"min"`
	Max      float64 `bson:"max" json:"max"`
}

type Risk struct {
	Description string `bson:"description" json:"description"`
	Impact      string `bson:"impact" json:"impact"`
	Mitigation  string `bson:"mitigation" json:"mitigation"`
}

type InternalNote struct {
	AuthorID  string    `bson:"author_id" json:"author_id"`
	Content   string    `bson:"content" json:"content"`
	CreatedAt time.Time `bson:"created_at" json:"created_at"`
}

func NewSpecification(projectID string) *Specification {
	now := time.Now()
	return &Specification{
		ID:        uuid.New().String(),
		ProjectID: projectID,
		Version:   1,
		Status:    SpecStatusDraft,
		CreatedAt: now,
		UpdatedAt: now,
	}
}

File: apps/server/src/domain/entity/specification.ts
import { ObjectId } from "mongodb";

export type SpecStatus =
  | "draft"
  | "generated"
  | "reviewed"
  | "approved"
  | "rejected";

export interface FeatureSpec {
  id: string;
  name: string;
  description: string;
  acceptanceCriteria: string[];
  priority: "low" | "medium" | "high" | "critical";
  estimatedEffort: string;
}

export interface RolePermission {
  role: string;
  permissions: string[];
  description: string;
}

export interface TechArchitecture {
  frontend: string;
  backend: string;
  database: string;
  infrastructure: string;
  diagram?: string;
  notes: string[];
}

export interface IntegrationReq {
  name: string;
  type: string;
  description: string;
  apiEndpoints: string[];
  authMethod: string;
}

export interface PhaseEstimate {
  phase: string;
  description: string;
  durationWeeks: number;
  dependencies: string[];
}

export interface TimelineEstimate {
  totalWeeks: number;
  phases: PhaseEstimate[];
}

export interface CostBreakdown {
  category: string;
  description: string;
  amount: number;
  currency: string;
}

export interface CostEstimate {
  total: number;
  currency: string;
  breakdown: CostBreakdown[];
}

export interface Risk {
  id: string;
  description: string;
  likelihood: "low" | "medium" | "high";
  impact: "low" | "medium" | "high";
  mitigation: string;
}

export interface InternalNote {
  id: string;
  authorId: string;
  content: string;
  createdAt: Date;
}

export interface Specification {
  id: string;
  projectId: string;
  status: SpecStatus;
  version: number;
  overview: string;
  objectives: string[];
  featureBreakdown: FeatureSpec[];
  userRolesPermissions: RolePermission[];
  technicalArchitecture: TechArchitecture;
  integrationReqs: IntegrationReq[];
  timelineEstimate: TimelineEstimate;
  costEstimate: CostEstimate;
  assumptions: string[];
  risks: Risk[];
  internalNotes: InternalNote[];
  generatedBy?: string;
  approvedBy?: string;
  approvedAt?: Date;
  createdAt: Date;
  updatedAt: Date;
}

export interface CreateSpecificationInput {
  projectId: string;
  overview?: string;
  objectives?: string[];
}

export function createSpecification(
  input: CreateSpecificationInput,
): Specification {
  const now = new Date();
  return {
    id: new ObjectId().toHexString(),
    projectId: input.projectId,
    status: "draft",
    version: 1,
    overview: input.overview ?? "",
    objectives: input.objectives ?? [],
    featureBreakdown: [],
    userRolesPermissions: [],
    technicalArchitecture: {
      frontend: "",
      backend: "",
      database: "",
      infrastructure: "",
      notes: [],
    },
    integrationReqs: [],
    timelineEstimate: { totalWeeks: 0, phases: [] },
    costEstimate: { total: 0, currency: "USD", breakdown: [] },
    assumptions: [],
    risks: [],
    internalNotes: [],
    createdAt: now,
    updatedAt: now,
  };
}
